evalml/pipelines/components/transformers/encoders/onehot_encoder.py

```diff
@@ -180,10 +179,37 @@ def categories(self, feature_name):
             raise ValueError(f'Feature "{feature_name}" was not provided to one-hot encoder as a training feature')
         return self._encoder.categories_[index]
 
+    @staticmethod
+    def _make_name_unique(name, seen_before):
+        """Helper to make the name unique."""
+        i = 1
+        while name in seen_before:
+            name = f"{name}_{i}"
+            i += 1
+        return name
+
     def get_feature_names(self):
         """Return feature names for the input features after fitting.
 
         Returns:
             np.array: The feature names after encoding, provided in the same order as input_features.
         """
-        return self._encoder.get_feature_names(self.features_to_encode)
+        unique_names = []
```

Comment:
Tested this on a df with two columns with 10,000 unique categories and top_n set to None. `timeit` reports a mean time of 5.75 ms which I think is acceptable. Especially since we usually pass in top_n and name clashes are pretty rare in my experience.